### PR TITLE
[SDKIOS-1024] Adopting SDK develop branch

### PR DESCRIPTION
**Pull request by szymon-onewelcome**

### CODE REVIEW

File: OneWelcomeExampleApp/ErrorMappings/AppToWebSingleSignOnErrorDomainMapping.swift [hunk 1]
```diff
@@ -5,7 +5,7 @@ class AppToWebSingleSignOnErrorDomainMapping {
         let title = "App to web error"
         
         switch error.code {
-        case ONGAppToWebSingleSignOnError.userNotAuthenticated.rawValue:
+        case AppToWebSingleSignOnError.userNotAuthenticated.rawValue:
```
Comment (kfa-onewelcome) [inline OneWelcomeExampleApp/ErrorMappings/AppToWebSingleSignOnErrorDomainMapping.swift:8]: and just ` .userNotAuthenticated`

File: OneWelcomeExampleApp/ErrorMappings/AuthenticatorDeregistrationErrorMapping.swift [hunk 2]
```diff
@@ -20,11 +20,11 @@ class AuthenticatorDeregistrationErrorMapping {
     
     func mapError(_ error: Error) -> AppError {
         switch error.code {
```
Comment (kfa-onewelcome) [inline OneWelcomeExampleApp/ErrorMappings/AuthenticatorDeregistrationErrorMapping.swift:22]: switch AuthenticatorDeregistrationError(rawValue: error.code)

File: OneWelcomeExampleApp/ErrorMappings/AuthenticatorRegistrationErrorDomainMapping.swift [hunk 3]
```diff
@@ -17,7 +17,7 @@ import UIKit
 
 class AuthenticatorRegistrationErrorDomainMapping {
     func mapError(_ error: Error, customInfo: CustomInfo?) -> AppError {
-        if let customInfo = customInfo, error.code == ONGAuthenticatorRegistrationError.customAuthenticatorFailure.rawValue {
```
Comment (kfa-onewelcome) [inline OneWelcomeExampleApp/ErrorMappings/AuthenticatorRegistrationErrorDomainMapping.swift:20]: you can make it switch

File: OneWelcomeExampleApp/ErrorMappings/AuthenticatorRegistrationErrorDomainMapping.swift [hunk 4]
```diff
@@ -30,15 +30,15 @@ private extension AuthenticatorRegistrationErrorDomainMapping {
 
     func mapError(_ error: Error) -> AppError {
         switch error.code {
```
Comment (kfa-onewelcome) [inline OneWelcomeExampleApp/ErrorMappings/AuthenticatorRegistrationErrorDomainMapping.swift:32]: switch AuthenticatorRegistrationError(rawValue: error.code)

File: OneWelcomeExampleApp/ErrorMappings/ErrorMapper.swift [hunk 5]
```diff
@@ -16,23 +16,23 @@
 class ErrorMapper {
     func mapError(_ error: Error, pinChallenge: PinChallenge? = nil, customInfo: CustomInfo? = nil) -> AppError {
         switch error.domain {
```
Comment (kfa-onewelcome) [inline OneWelcomeExampleApp/ErrorMappings/ErrorMapper.swift:18]: switch ErrorDomains(rawValue: error.domain)

File: OneWelcomeExampleApp/ErrorMappings/FetchImplicitResourceErrorDomainMapping.swift [hunk 6]
```diff
@@ -20,7 +20,7 @@ class FetchImplicitResourceErrorDomainMapping: NSObject {
         let title = "Fetching implicit resource error"
 
         switch error.code {
```
Comment (kfa-onewelcome) [inline OneWelcomeExampleApp/ErrorMappings/FetchImplicitResourceErrorDomainMapping.swift:22]: switch FetchResourceImplicitlyError(rawValue: error.code)

File: OneWelcomeExampleApp/ErrorMappings/FetchResourceErrorMapping.swift [hunk 7]
```diff
@@ -3,7 +3,7 @@
 class FetchResourceErrorDomainMapping {
     func mapError(_ error: Error) -> AppError {
         switch error.code {
```
Comment (kfa-onewelcome) [inline OneWelcomeExampleApp/ErrorMappings/FetchResourceErrorMapping.swift:5]: switch FetchResourceError(rawValue: error.code)

File: OneWelcomeExampleApp/ErrorMappings/GenericErrorDomainMapping.swift [hunk 8]
```diff
@@ -16,17 +16,17 @@
 class GenericErrorDomainMapping {
     func mapError(_ error: Error) -> AppError {
         switch error.code {
```
Comment (kfa-onewelcome) [inline OneWelcomeExampleApp/ErrorMappings/GenericErrorDomainMapping.swift:18]: switch GenericError(rawValue: error.code)

File: OneWelcomeExampleApp/ErrorMappings/MobileAuthEnrollmentErrorDomainMapping.swift [hunk 9]
```diff
@@ -18,27 +18,27 @@ class MobileAuthEnrollmentErrorDomainMapping {
         let title = "Mobile auth enrollment error"
 
         switch error.code {
```
Comment (kfa-onewelcome) [inline OneWelcomeExampleApp/ErrorMappings/MobileAuthEnrollmentErrorDomainMapping.swift:20]: switch MobileAuthEnrollmentError(rawValue: error.code)

File: OneWelcomeExampleApp/Interactors/RegisterUserInteractor.swift [hunk 10]
```diff
@@ -131,6 +131,9 @@ extension RegisterUserInteractor: RegistrationDelegate {
         if challenge.identityProvider.identifier == "2-way-otp-api" {
             challenge.sender.respond(with: nil, to: challenge)
         }
+        if challenge.identityProvider.identifier == "jp_na100procent" {
+            challenge.sender.respond(with: nil, to: challenge)
+        }
```
[the remark this answers is not in the document]
Comment (szymon-onewelcome) [reply OneWelcomeExampleApp/Interactors/RegisterUserInteractor.swift:136]: God catch, leftover from testing custom idp :trollface:

File: OneWelcomeExampleApp/ErrorMappings/ErrorMapper.swift (hunk 5 above)
Comment (szymon-onewelcome) [reply OneWelcomeExampleApp/ErrorMappings/ErrorMapper.swift:18]: `ErrorDomains` is not yet a propper enum
```
// TODO: To be replaced with proper enum in pure Swift
// swiftlint:disable line_length
@objcMembers
public class ErrorDomains: NSObject {

**Review by kfa-onewelcome — CHANGES_REQUESTED** [review]

File: OneWelcomeExampleApp/Interactors/LoginInteractor.swift [hunk 11]
```diff
@@ -40,7 +40,7 @@ class LoginInteractor: NSObject, LoginInteractorProtocol {
     }
 
     fileprivate func mapErrorFromChallenge(_ challenge: PinChallenge) {
-        if let error = challenge.error, error.code != ONGAuthenticationError.touchIDAuthenticatorFailure.rawValue {
+        if let error = challenge.error, error.code != AuthenticationError.touchIDAuthenticatorFailure.rawValue {
```
[the remark this answers is not in the document]
Comment (szymon-onewelcome) [reply OneWelcomeExampleApp/Interactors/LoginInteractor.swift:43]: I did most of the cases, this one I missed.

File: OneWelcomeExampleApp/Interactors/LoginInteractor.swift [hunk 12]
```diff
@@ -98,7 +97,7 @@ extension LoginInteractor: AuthenticationDelegate {
     }
 
     func userClient(_ userClient: UserClient, didFailToAuthenticateUser userProfile: UserProfile, authenticator: Authenticator, error: Error) {
-        if error.code == ONGGenericError.actionCancelled.rawValue {
```
[the remark this answers is not in the document]
Comment (szymon-onewelcome) [reply OneWelcomeExampleApp/Interactors/LoginInteractor.swift:101]: Same here. Let me know if you see some more.

File: README.md [hunk 13]
```diff
@@ -9,11 +9,11 @@ example app.
 ### Setup access to the Onegini Cocoapods repository
 The Example app includes the Onegini SDK as a Cocoapod. In order to let Cocoapods download it you need to setup your account details so the SDK can be
 automatically downloaded:
-1. Make sure that you have access to the Onegini Artifactory repository (https://repo.onegini.com). If not please follow first step of [App developer quickstart](https://docs.onegini.com/app-developer-quickstart.html).
-2. Follow [Setting up the project guide](https://docs.onegini.com/ios-sdk/topics/setting-up-the-project.html#cocoapods) in the Onegini SDK documentation for
-instructions on configuring access to the Onegini Cocoapods repository.
+1. Make sure that you have access to the OneWelcome Artifactory repository (https://repo.onewelcome.com). If not please follow first step of [App developer quickstart](https://docs.onegini.com/app-developer-quickstart.html).
```
[the remark this answers is not in the document]
Comment (szymon-onewelcome) [reply README.md:12]: What is the current link to docs under onewelcome domain? Hitting this one gives me `https://docs-single-tenant.onegini.com/app-developer-quickstart.html`?

**Review by kfa-onewelcome — CHANGES_REQUESTED** [review]
I'm not able to install clean pods. Please remove the project, download from GitHub , deintegrate pods and install pods again.

File: Podfile [hunk 14]
```diff
@@ -6,7 +6,7 @@ use_frameworks!
 plugin 'cocoapods-art', :sources => ['onegini']
 
 def oneginiSDKiOS
-   pod 'OneginiSDKiOS', '~> 12.2.0'
+    pod 'OneginiSDKiOS', :git => 'git@github.com:onewelcome/sdk-ios.git', :branch => 'develop'
```
Comment (kfa-onewelcome) [inline Podfile:9]: This is not working, that's why I'm not able to build the project

**Review by kfa-onewelcome — CHANGES_REQUESTED** [review]

**Review by kfa-onewelcome — CHANGES_REQUESTED** [review]
The new finding: the example app crashes when registers the dummy IDP, please also check other scenarios w/out payload encryption and on Access (not TS)

File: Podfile (hunk 14 above)
Comment (szymon-onewelcome) [reply Podfile:9]: I've tried hard. The problem is that module map. If this is for developemnt how about putting here sth like 
`pod 'OneginiSDKiOS', :path => 'path/to/local/sdk/repo/on/develop/branch'` ?

**Review by kfa-onewelcome — CHANGES_REQUESTED** [review]
Did you test different flows? 

I have another issue with the current state: when you register a new profile and logout and try to login and provide a PIN wrong once(!) you are deregistered.

**Comment by kfa-onewelcome** [comment]
Could please also merge changes from master do develop (stateless)

**Comment by szymon-onewelcome** [comment]
> Did you test different flows?
> 
> I have another issue with the current state: when you register a new profile and logout and try to login and provide a PIN wrong once(!) you are deregistered.

Even though pinChallenge has some remaining pin attempts but associated user profile stands it's 0.
```
(lldb) po pinChallenge.remainingFailureCount
3
(lldb) po pinChallenge.userProfile.remainingPinAttempts
0
```

So we this is what we get from TS:
```
Error Domain=com.onegini.GenericErrorDomain Code=9003 "The user account is deregistered from the device. The user supplied the wrong PIN for too many times. All local data associated with the user profile has been removed." UserInfo={NSLocalizedDescription=The user account is deregistered from the device. The user supplied the wrong PIN for too many times. All local data associated with the user profile has been removed.}
```

I'm investigation this incoherente, I guess some data is lost when swift/objc instances are created back and forth.

**Comment by szymon-onewelcome** [comment]
> Did you test different flows?
> 
> I have another issue with the current state: when you register a new profile and logout and try to login and provide a PIN wrong once(!) you are deregistered.

Bug in SDK, PR to fix is here:
https://github.com/onewelcome/sdk-ios/pull/281/files

**Review by kfa-onewelcome — COMMENTED** [review]
Can you check why I'm not able to register for mobile authentication? We get 10000 error?

**Review by kfa-onewelcome — CHANGES_REQUESTED** [review]
I was able to enroll for mobile auth on master (EXAMPLEIOS-88)

pls check

File: Configuration/Production/OneginiConfigModel.m [hunk 15]
```diff
@@ -2,28 +2,28 @@
...
 + (NSDictionary *)configuration
 {
     return @{
              @"ONGAppIdentifier" : @"ExampleApp",
              @"ONGAppPlatform" : @"ios",
-             @"ONGAppVersion" : @"6.0.0",
-             @"ONGAppBaseURL" : @"https://demo-msp.onegini.com",
-             @"ONGResourceBaseURL" : @"https://demo-msp.onegini.com/resources/",
+             @"ONGAppVersion" : @"6.2.0",
+             @"ONGAppBaseURL" : @"https://mobile-security-proxy.onegini.com",
```
Comment (kfa-onewelcome) [inline Configuration/Production/OneginiConfigModel.m:18]: I would set `.test.` for develop and prod only for prod

Comment (szymon-onewelcome) [reply Configuration/Production/OneginiConfigModel.m:18]: I've changed config for prod only, the same as you have in `Example-88`.

Comment (kfa-onewelcome) [reply Configuration/Production/OneginiConfigModel.m:18]: aww right, this is a Production config

**Review by kfa-onewelcome — APPROVED** [review]

**Pull request merged**